proposals/p2188.md

```diff
@@ -0,0 +1,1689 @@
+        the new value is not handled.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Pattern syntax makes it easier to match complex values.
+    -   Modeling pattern syntax after expressions eases the burden of learning a
+        new sub-language for pattern-matching: patterns are an extension of
+        expressions, and expressions are a special case of patterns.
+    -   Requiring exhaustiveness for matches makes control flow easier to
+        understand as there is never a value for which the `match` skips all
+        cases.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   The rules for matching a templated value can be used to replace
+        `if constexpr` in many cases.
+
+## Alternatives considered
+
+### Shorthand for `auto`
+
+We could provide a shorter syntax for `name: auto`.
+[Proposal #851](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0851.md#elide-the-type-instead-of-using-auto)
+considered the following shorthands and decided against using them:
+
+```
+var n: _ = init;
+var n = init;
+```
+
+A novel suggestion that avoids some of the disadvantages of those syntaxes would
+be to use:
+
+```
+var n:= init;
+```
+
+Advantages:
+
+-   Shorter syntax for variables with a deduced type.
+-   Potentially allows removal of the `auto` keyword.
+
+Disadvantages:
+
+-   Appears to introduce a `:=` syntax, but that only arises in cases where an
+    initializer immediately follows the name.
+    -   Cases such as `var (a:, b:) = my_pair;` would either be invalid or would
+        not use the `:=` syntax.
+    -   If we accept such cases, there is a risk of grammar ambiguities.
+    -   If we reject such cases, we may still want to keep `auto` around for
+        them, creating inconsistency.
+-   Not a complete replacement for `auto` if we want to also allow things like
+    `v: Vector(auto)`. C++ doesn't allow the equivalent syntax currently, but it
+    was part of the Concepts TS and seems likely to return at some point.
+-   No syntactic difference between accidentally omitting a type entirely and
+    requesting type deduction. However, the mistake of omitting a type but
+    retaining the `:` seems unlikely, and the `:` followed by the absence of a
+    type is a signal that something is happening, so this seems to be less of a
+    concern than for the `var n = init;` syntax.
+
+See discussion topics
+[1](https://github.com/carbon-language/carbon-lang/discussions/1495) and
+[2](https://github.com/carbon-language/carbon-lang/discussions/1988).
+
+### Struct pattern syntax
+
+We could omit the `, _` syntax. This would simplify struct patterns, but at the
+cost of removing a feature that can be useful for reducing verbosity and making
+library evolution easier.
```

Comment:
I'm not fully convinced that we should add this yet TBH.

I wonder if we'll be able to address this usecase *with* variadics, and with tools that also address discarding an unspecified number of tuple elements.

I lean a little bit towards maybe documenting this option but waiting to see if we instead can solve this as part of variadic patterns more generally. What do other folks think here?

The flip side is that I think this is a minor feature and easily *removed* if we find that variadics obviate it, so I also don't think this is a big deal either way.

Reply:
Note that struct fields aren't expressions or values, so the ability to create and match packs of struct fields would be a separate, bespoke feature, not something that would emerge automatically from a more general variadics design. I don't think we've discussed having such a feature, and it's not clear to me if there are any other use cases for it (maybe metaprogramming?). So my inclination would be to keep `, _` for now, and then replace it if and when we discover we want packs-of-struct-fields for other reasons.